ui/src/components/build/CreateBuildConfigDialog.js

```diff
@@ -0,0 +1,189 @@
+        ...this.props.config
+      }
+    };
+
+    this.configState = new SubState(this, 'config', configValidation, [KEY_CR_NAME]);
+    this.buildState = new SubState(this, 'config.build', buildValidation);
+    this.basicAuthState = new SubState(this, 'config.source.basicAuth', basicAuthValidation, [
+      KEY_CR_BASIC_AUTH_NAME,
+      KEY_CR_BASIC_AUTH_PASSWORD,
+      KEY_CR_BASIC_AUTH_USERNAME
+    ]);
+    this.sshAuthState = new SubState(this, 'config.source.sshAuth', sshAuthValidation, [
+      KEY_CR_SSH_AUTH_NAME,
+      KEY_CR_SSH_PRIVATE_KEY
+    ]);
+    this.sourceState = new SubState(this, 'config.source', sourceValidation, [KEY_CR_SOURCE_GITURL]);
+    this.androidCredentialsState = new SubState(this, 'config.build.androidCredentials', androidCredentialsValidation, [
+      KEY_CR_BUILD_ANDROID_CREDENTIALS_NAME,
+      KEY_CR_BUILD_ANDROID_CREDENTIALS_KEYSTORE,
+      KEY_CR_BUILD_ANDROID_CREDENTIALS_KEYSTORE_ALIAS,
+      KEY_CR_BUILD_ANDROID_CREDENTIALS_KEYSTORE_PASSWORD
+    ]);
+    this.iOSCredentialsState = new SubState(this, 'config.build.iosCredentials', iOSCredentialsValidation, [
+      KEY_CR_BUILD_IOS_CREDENTIALS,
+      KEY_CR_BUILD_IOS_CREDENTIALS_DEVELOPER_PROFILE,
+      KEY_CR_BUILD_IOS_CREDENTIALS_NAME
+    ]);
+  }
+
+  componentDidUpdate(prevProps) {
+    this.validate();
+    const { initialConfig: prevConfig = {} } = prevProps;
+    const { initialConfig = {} } = this.props;
+    if (!isEqual(prevConfig, initialConfig)) {
+      this.configState.set(KEY_CR_CLIENT_ID, initialConfig[KEY_CR_CLIENT_ID]);
+      this.configState.set(KEY_CR_CLIENT_TYPE, initialConfig[KEY_CR_CLIENT_TYPE]);
+      switch (initialConfig[KEY_CR_CLIENT_TYPE]) {
+        case CLIENT_TYPE_ANDROID:
+          this.buildState.set(KEY_CR_BUILD_PLATFORM, BUILD_PLATFORM_ANDROID);
+          this.setState({ [KEY_HIDE_PLATDORM]: true });
+          break;
+        case CLIENT_TYPE_IOS:
+          this.buildState.set(KEY_CR_BUILD_PLATFORM, BUILD_PLATFORM_IOS);
+          this.setState({ [KEY_HIDE_PLATDORM]: true });
+          break;
+        default:
+          this.setState({ [KEY_HIDE_PLATDORM]: false });
+      }
+    }
+  }
+
+  validate() {
+    let validVal = this.configState.isAllValid() && this.buildState.isAllValid() && this.sourceState.isAllValid();
+    if (this.sourceState.get(KEY_CR_SOURCE_AUTH_TYPE) === BUILD_AUTH_TYPE_BASIC)
+      validVal = validVal && this.basicAuthState.isAllValid();
+    if (this.sourceState.get(KEY_CR_SOURCE_AUTH_TYPE) === BUILD_AUTH_TYPE_SSH)
+      validVal = validVal && this.sshAuthState.isAllValid();
+    if (this.state.valid !== validVal) this.setState({ valid: validVal });
+  }
+
+  onSaveBuildConfig = () => {
+    const { config } = this.state;
+    const { onSave, update = false, initialConfig } = this.props;
+    onSave && onSave({ update, config: { ...initialConfig, ...config } });
```

Comment:
Is `{ ...initialConfig, ...config }` needed? Should not `this.state.config` be initializes with `this.props.config`/`this.props.initialConfig` in constructor? What is difference between `this.props.config` and `this.props.initialConfig`?

Reply:
removed